### ISSUE TITLE
PyG GNNexplainer generating different explanations

### ISSUE THREAD

**Issue by BILLYLIAOWEI** [rusty1s/pytorch_geometric#2034]
## ❓ Questions & Help

I'm using GCN to classify the node on datasets "KarateClub", then I using GNNExplainer to explain node 12. However, when I explain the node twice, GNNExplainer gives me two different subgraph, and them have different node_feat_mask&edge_mask.
I‘m so confused about the different explanations generated from the same trained model.

### run code on jupyter notebook 
_cell one_
`
import torch
from torch_geometric.datasets import *
import torch.nn.functional as F
from torch_geometric.nn import GCNConv,GNNExplainer
import networkx as nx
import matplotlib.pyplot as plt
from torch_geometric.datasets import *

dataset = KarateClub()
#print(dataset[0])

class Net(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = GCNConv(dataset.num_node_features, 16)
        self.conv2 = GCNConv(16, dataset.num_classes)
        pass

    def forward(self,x,edge_index):
        #x, edge_index = data.x, data.edge_index

        x = self.conv1(x, edge_index)
        x = F.relu(x)
        x = F.dropout(x, training=self.training)
        x = self.conv2(x, edge_index)

        return F.log_softmax(x, dim=1)
    pass
pass

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model = Net().to(device)
data = dataset[0].to(device)
optimizer = torch.optim.Adam(model.parameters(), lr=0.01)
x, edge_index = data.x, data.edge_index

for epoch in range(61):
    model.train()
    optimizer.zero_grad()
    out = model(x,edge_index) #log_logits = model(x, edge_index)
    loss = F.nll_loss(out, data.y) #loss = F.nll_loss(log_logits[data.train_mask], data.y[data.train_mask])
    loss.backward()
    optimizer.step()

    #print('Epoch {} | Loss: {:.4f}'.format(epoch,loss.item()))
    model.eval()
    _, pred = model(x,edge_index).max(dim=1)
    #print(pred)
    #print(pred.eq(data.y))
    correct = int(pred.eq(data.y).sum().item())
    acc = correct / int(data.x.sum())
    #print('Accuracy:{:.4f}'.format(acc))
    print('Epoch {} | Loss: {:.4f}'.format(epoch,loss.item())+' | Accuracy:{:.4f}'.format(acc))
    #if epoch%20==0:
        #plot_graph(pred,G,epoch)
        #plt.show()

    pass
pass

explainer = GNNExplainer(model, epochs=60)
node_idx = 12
node_feat_mask, edge_mask = explainer.explain_node(node_idx, x, edge_index)
ax, G = explainer.visualize_subgraph(node_idx, edge_index, edge_mask, y=data.y)
plt.show()
`
### _cell two_
`
explainer = GNNExplainer(model, epochs=60)
node_idx = 12
node_feat_mask, edge_mask = explainer.explain_node(node_idx, x, edge_index)
ax, G = explainer.visualize_subgraph(node_idx, edge_index, edge_mask, y=data.y)
plt.show()
`

**Comment by rusty1s** [comment]
Yes, I noticed this myself, but it is kind of reasonable given that `GNNExplainer` is a gradient descent optimization procedure. In the end, there may exists multiple explanations that do not change the outcome predictions that much. I believe you can get much stabler performance when tuning hyperparameters, such as `lr`, or the coefficients given in `GNNExpaliner.coeffs`.

**Comment by joaquincabezas** [comment]
Hello @BILLYLIAOWEI 

I have experienced this instability with the method before. I found it while explaining nodes with small neighborhoods. We have discussed it in this thread: https://github.com/rusty1s/pytorch_geometric/issues/1985

Please take a look at the code provided and check if it solves your issue. We are considering putting a note, so it will really help if others test this proposal too. Thanks!
